src/Microsoft.Restier.Core/ApiConfiguration.cs

```diff
@@ -65,5 +68,46 @@ public IServiceProvider ServiceProvider
         {
             return this.serviceProvider.GetService<T>();
         }
+
+        internal TaskCompletionSource<IEdmModel> CompeteModelGeneration(out Task<IEdmModel> running)
+        {
+            var source = new TaskCompletionSource<IEdmModel>(TaskCreationOptions.AttachedToParent);
+            var exist = Interlocked.CompareExchange(ref modelTask, source.Task, null);
+            if (exist != null)
+            {
+                running = exist;
+                source.SetCanceled();
+                return null;
+            }
+
+            source.Task.ContinueWith(
```

Comment:
Oh yeah, I got your thought. Yes another thread would continue on the original task source task, not the appended one, so it could possibly run before EdmModel is set, but as you pointed out this won't hurt, since anyway the resulting IEdmModel is retrieved.